### PR TITLE
feat(p2p): support IPv6

### PR DESCRIPTION
**Pull request by luislhl**
### Motivation

We want to support IPv6 connections between fullnodes and support exposing their API through IPv6

### Acceptance Criteria

- We should be able to listen the API, the Stratum and the P2P port in IPv6 interfaces. This will be activated by a new CLI argument deemed initially as experimental: `--x-enable-ipv6`
- We will also have a `--x-disable-ipv4`, to make nodes stop connecting to IPv4 endpoints. For now this will be used mostly for tests though, to force a fullnode to behave as IPv6-only.
- The peers should relay IPv6 endpoints only to other peers that also have the IPv6 capability. Relaying them to old fullnodes would cause errors when they try to parse IPv6.
- For simplicity, this PR only implements support for IPv6 addresses in endpoints, not domains that contain AAAA records

### TODO

- [ ] After merging, we should think about how to have permanent nodes that are dualstack and ipv6-only so we always test this is working

### Checklist

- [ ] If you are requesting a merge into `master`, confirm this code is production-ready and can be included in future releases as soon as it gets merged

### CODE REVIEW

File: hathor/metrics.py
Comment (jansegre) [inline hathor/metrics.py:1]: I'm assuming this change will disappear after the last #1145 update.

File: hathor/p2p/peer.py
Comment (jansegre) [inline hathor/p2p/peer.py:1]: This module has been refactored a lot and the PR was just merged. Let me know if you need help rebasing.

File: hathor/p2p/states/peer_id.py [hunk 1]
```diff
@@ -68,10 +74,16 @@ def send_peer_id(self) -> None:
         """
         protocol = self.protocol
         my_peer = protocol.my_peer
+
+        if not self.should_relay_ipv6_entrypoints:
+            entrypoints_as_str = my_peer.ipv4_entrypoints_as_str()
+        else:
+            entrypoints_as_str = my_peer.entrypoints_as_str()
```
Comment (jansegre) [inline hathor/p2p/states/peer_id.py:81]: Should we restrict `my_peer.entrypoints_as_str()` if IPv4 is disabled? It could make sense to at least issue warnings if there are IPv4 addresses in `my_peer.entrypoints`.

File: hathor/p2p/states/peer_id.py [hunk 2]
```diff
@@ -89,7 +101,6 @@ async def handle_peer_id(self, payload: str) -> None:
         data = json_loads(payload)
 
         peer = Peer.create_from_json(data)
-        peer.validate()
```
Comment (jansegre) [inline hathor/p2p/states/peer_id.py:92]: Why was this removed? Either way this will conflict with the latest refactor.

**Review by jansegre — APPROVED** [review]

File: hathor/p2p/states/peer_id.py (hunk 2 above)
Comment (luislhl) [reply hathor/p2p/states/peer_id.py:92]: Because I had added this validation to be done inside of `create_from_json`. It seems you did this too in the PR merged to master, so it still made sense to remove this while I was fixing conflicts.

File: hathor/metrics.py
Comment (luislhl) [reply hathor/metrics.py:1]: It did

File: hathor/p2p/peer.py
Comment (luislhl) [reply hathor/p2p/peer.py:1]: I've rebased it already. Can you check?

File: hathor/p2p/states/peer_id.py (hunk 1 above)
Comment (luislhl) [reply hathor/p2p/states/peer_id.py]: I did this way for simplicity. My objective with this check was to avoid relaying IPv6 to nodes that aren't ready to parse them.

All nodes are ready to parse IPv4, so it didn't seem necessary to limit them.

However, nodes that have IPv4 disabled will not try to connect to IPv4 entrypoints they receive.

But, thinking about your suggestion, isn't it better to let the IPv4 addresses be relayed even by nodes that have it disabled, so it could pass them forward to others that might use them? In some very specific (and unlikely) cases this could be useful, like if some node needs to first connect to some other that is IPv6-only as a bootstrap.

**Review by jansegre — APPROVED** [review]

File: hathor/p2p/states/peer_id.py (hunk 1 above)
Comment (jansegre) [reply hathor/p2p/states/peer_id.py]: I think it makes sense to try to propagate all IPs as much as possible (and just avoid sending IPv6 addresses to peers that can't parse them).

File: hathor/cli/nginx_config.py [hunk 3]
```diff
@@ -240,11 +240,12 @@ def generate_nginx_config(openapi: dict[str, Any], *, out_file: TextIO, rate_k:
 
     server_open = f'''
 upstream backend {{
-    server fullnode:8080;
+    server 127.0.0.1:8080;
 }}
 
 server {{
     listen 80;
+    listen [::]:80;
```
Comment (msbrogli) [inline hathor/cli/nginx_config.py:248]: Should we add an `--enable-ipv6` here too?

File: hathor/cli/run_node.py [hunk 4]
```diff
@@ -201,7 +206,11 @@ def prepare(self, *, register_resources: bool = True) -> None:
 
         if self._args.stratum:
             assert self.manager.stratum_factory is not None
-            self.reactor.listenTCP(self._args.stratum, self.manager.stratum_factory)
+
+            if self._args.x_enable_ipv6:
+                self.reactor.listenTCP(self._args.stratum, self.manager.stratum_factory, interface='::0')
+            else:
+                self.reactor.listenTCP(self._args.stratum, self.manager.stratum_factory)
```
Comment (msbrogli) [inline hathor/cli/run_node.py:213]: Shouldn't we listen in both interfaces (ipv4 and ipv6) when ipv6 is enabled?

File: hathor/cli/run_node.py [hunk 5]
```diff
@@ -201,7 +206,11 @@ def prepare(self, *, register_resources: bool = True) -> None:
 
         if self._args.stratum:
             assert self.manager.stratum_factory is not None
-            self.reactor.listenTCP(self._args.stratum, self.manager.stratum_factory)
+
+            if self._args.x_enable_ipv6:
+                self.reactor.listenTCP(self._args.stratum, self.manager.stratum_factory, interface='::0')
```
Comment (msbrogli) [inline hathor/cli/run_node.py:211]: Why `::0`? I understand this might be default but there should be a way to choose one specific interface.

File: hathor/cli/run_node.py [hunk 6]
```diff
@@ -216,7 +225,11 @@ def prepare(self, *, register_resources: bool = True) -> None:
             status_server = resources_builder.build()
             if self._args.status:
                 assert status_server is not None
-                self.reactor.listenTCP(self._args.status, status_server)
+
+                if self._args.x_enable_ipv6:
+                    self.reactor.listenTCP(self._args.status, status_server, interface='::0')
+                else:
+                    self.reactor.listenTCP(self._args.status, status_server)
```
Comment (msbrogli) [inline hathor/cli/run_node.py:232]: Same here. Shouldn't we listen in both ipv4 and ipv6 interfaces?

File: hathor/metrics.py
Comment (msbrogli) [inline hathor/metrics.py:1]: These changes seem unrelated to this PR.

File: hathor/p2p/entrypoint.py [hunk 7]
```diff
@@ -12,18 +12,21 @@
...
 from twisted.internet.address import IPv4Address, IPv6Address
-from twisted.internet.endpoints import TCP4ClientEndpoint
+from twisted.internet.endpoints import TCP4ClientEndpoint, TCP6ClientEndpoint
 from twisted.internet.interfaces import IStreamClientEndpoint
 from typing_extensions import Self
 
 from hathor.p2p.peer_id import PeerId
 from hathor.reactor import ReactorProtocol as Reactor
 
+IPV6_REGEX = re.compile(r'''^(([0-9a-fA-F]{1,4}:){7}([0-9a-fA-F]{1,4}|:)|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:))$''')  # noqa: E501
```
Comment (msbrogli) [inline hathor/p2p/entrypoint.py:28]: Can you add a docstring and some examples that match and do not match this regex, please?

File: hathor/p2p/entrypoint.py [hunk 8]
```diff
@@ -39,10 +42,14 @@ class Entrypoint:
     peer_id: PeerId | None = None
```
Comment (msbrogli) [inline hathor/p2p/entrypoint.py:42]: Should we add a `tcp6` protocol instead of trying to match a regex? Alternatively, we could use the `[host]` pattern for all ipv6 hosts.

File: hathor/p2p/manager.py [hunk 9]
```diff
@@ -580,7 +588,11 @@ def _update_whitelist_cb(self, body: bytes) -> None:
     def connect_to_if_not_connected(self, peer: UnverifiedPeer | PublicPeer, now: int) -> None:
         """ Attempts to connect if it is not connected to the peer.
         """
-        if not peer.info.entrypoints:
+        if not peer.info.entrypoints or (
```
Comment (msbrogli) [inline hathor/p2p/manager.py:591]: I didn't get it. What is this change doing?

File: hathor/p2p/manager.py [hunk 10]
```diff
@@ -592,7 +604,14 @@ def connect_to_if_not_connected(self, peer: UnverifiedPeer | PublicPeer, now: in
...
         if peer.info.can_retry(now):
-            self.connect_to(self.rng.choice(peer.info.entrypoints), peer)
+            if self.enable_ipv6 and not self.disable_ipv4:
+                self.connect_to(self.rng.choice(peer.info.entrypoints), peer)
+            elif self.enable_ipv6 and self.disable_ipv4:
+                self.connect_to(self.rng.choice(peer.info.ipv6_entrypoints), peer)
+            elif not self.enable_ipv6 and not self.disable_ipv4:
+                self.connect_to(self.rng.choice(peer.info.ipv4_entrypoints), peer)
+            else:
+                raise ValueError('IPv4 is disabled and IPv6 is not enabled')
```
Comment (msbrogli) [inline hathor/p2p/manager.py:614]: You can use `assert_never()` here.

File: hathor/p2p/peer.py [hunk 11]
```diff
@@ -122,6 +122,20 @@ class PeerInfo:
     flags: set[str] = field(default_factory=set)
     _settings: HathorSettings = field(default_factory=get_global_settings, repr=False)
 
+    @property
+    def ipv4_entrypoints(self) -> list[Entrypoint]:
```
Comment (msbrogli) [inline hathor/p2p/peer.py:126]: I'd rather rename it to `get_ipv4_only_entrypoints()`.

File: hathor/p2p/peer.py [hunk 12]
```diff
@@ -122,6 +122,20 @@ class PeerInfo:
...
+        return list(filter(lambda e: not e.is_ipv6(), self.entrypoints))
+
+    @property
+    def ipv6_entrypoints(self) -> list[Entrypoint]:
```
Comment (msbrogli) [inline hathor/p2p/peer.py:130]: I'd rather rename it to `get_ipv6_only_entrypoints()`.

File: hathor/p2p/peer.py [hunk 13]
```diff
@@ -226,14 +240,19 @@ class UnverifiedPeer:
     id: PeerId
     info: PeerInfo = field(default_factory=PeerInfo)
 
-    def to_json(self) -> dict[str, Any]:
+    def to_json(self, only_ipv4_entrypoints: bool = True) -> dict[str, Any]:
```
Comment (msbrogli) [inline hathor/p2p/peer.py:243]: Why don't we add all entrypoints regardless of what's enabled/disabled?

**Review by msbrogli — CHANGES_REQUESTED** [review]

File: hathor/cli/nginx_config.py (hunk 3 above)
Comment (luislhl) [reply hathor/cli/nginx_config.py:248]: Actually, now that you mentioned, I'm wondering whether we actually need to add this here.

I had added it mostly to test IPv6 support in the API. But in our deploys, we won't really need it, because the API is server through a Load Balancer. The LB will still talk IPv4 with Nginx.

I think we could even remove it and it would make no difference.

However, I guess there is no harm in leaving it and always making it listen in the IPv6 interface, is there? The image would be ready for anyone that needs it eventually over IPv6, for any reason. And if it's not needed, it just won't be used.

Another thing is that we would be creating more variations of the Nginx Docker image that we generate. We already have the `latest` and `no-rate-limit-latest`.

We would probably need to generate additionally the `latest-with-ipv6` and `no-rate-limit-latest-with-ipv6`.

I would rather avoid these variations unless we have a good reason to not always listen on IPv6 in Nginx.

File: hathor/cli/run_node.py (hunk 4 above)
Comment (luislhl) [reply hathor/cli/run_node.py:213]: Initially I had this impression too. But it turns out Linux by default will map IPv4 to IPv6, and listening on `::0` is enough to handle IPv4 traffic as well.

I was actually getting an error when trying to listen on IPv4 simultaneously on the same port.

You can reproduce this behavior by running a simple server like this listening on `::`:

```
python3 -m http.server --bind :: 8080
```

And then connecting to it both over IPv6 or IPv4:

```
nc 127.0.0.1 8080
nc ::1 8080
```

There is a kernel paramenter to control this behavior, but the default is to translate IPv4 to IPv6 when bound to `::0`: https://sysctl-explorer.net/net/ipv6/bindv6only/

File: hathor/cli/run_node.py (hunk 5 above)
Comment (luislhl) [reply hathor/cli/run_node.py:211]: This is a good point. I'll add new parameters to allow changing this default.

File: hathor/cli/run_node.py (hunk 6 above)
Comment (luislhl) [reply hathor/cli/run_node.py:232]: Answered in the other thread.

File: hathor/p2p/peer.py (hunk 13 above)
Comment (luislhl) [reply hathor/p2p/peer.py:243]: This method was being used to build the JSON that is sent to other peers in `ReadyState.send_peers`.

I needed to avoid sending IPv6 endpoints to peers that can't handle it.

I could also just build the response manually there as an alternative, so I don't have to use this `.to_json` method. What do you think?

File: hathor/p2p/manager.py (hunk 10 above)
Comment (luislhl) [reply hathor/p2p/manager.py:614]: This was introduced in Python 3.11, right? 

We are still supporting 3.10, AFAIK, so I guess we should avoid this?

File: hathor/p2p/manager.py (hunk 9 above)
Comment (luislhl) [reply hathor/p2p/manager.py:591]: This code tries to remove from the list of peers the ones that disconnected from us and to which we surely won't be able to connect anymore.

Previously, to detect this, we only had to check whether they had no entrypoints.

But now, with IPv6, new cases could happen that wouldn't be caught:
- We don't have IPv6 enabled, and the peer only has IPv6 entrypoints (no IPv4)
- We have IPv4 disabled, and the peer only has IPv4 entrypoints (no IPv6)

It seemed like we would want to remove them from the list as well.

Does it make sense?

File: hathor/p2p/peer.py (hunk 11 above)
Comment (luislhl) [reply hathor/p2p/peer.py:126]: Will do

File: hathor/p2p/peer.py (hunk 12 above)
Comment (luislhl) [reply hathor/p2p/peer.py:130]: Will do

File: hathor/p2p/entrypoint.py (hunk 7 above)
Comment (luislhl) [reply hathor/p2p/entrypoint.py:28]: Will do

File: hathor/p2p/entrypoint.py (hunk 8 above)
Comment (luislhl) [reply hathor/p2p/entrypoint.py:42]: This is a good question, I thought about this at some point.

I chose not to rely so much in user input for this and just detect whether an IPv6 was provided.

Otherwise, someone could wrongly use `tcp` instead of `tcp6` as an entrypoint, and it would be relayed as if it was an IPv4 to others, that would error out while trying to parse it, or at least keep trying to connect to it and getting timeouts.

If we do start using `tcp6`, I think it would be better to keep the regex anyway for validation purposes to avoid the scenario above.

What do you think?

One of the advantages of using `tcp6` I thought about initially was that it could be used to also inform about DNS names holding AAAA records. For instance, `tcp6://abc.xyz` would be a way for the user to inform that this domain should be treated as IPv6. This something not supported in the current implementation with the Regex.

We would still have the validation issue, though, which is even harder in this case because it needs a DNS query.

File: hathor/metrics.py
Comment (luislhl) [reply hathor/metrics.py:1]: Removed

File: hathor/p2p/entrypoint.py (hunk 7 above)
Comment (luislhl) [reply hathor/p2p/entrypoint.py:57]: Done

File: hathor/p2p/peer.py (hunk 11 above)
Comment (luislhl) [reply hathor/p2p/peer.py]: Done

File: hathor/p2p/peer.py (hunk 12 above)
Comment (luislhl) [reply hathor/p2p/peer.py]: Done

File: hathor/cli/run_node.py (hunk 5 above)
Comment (luislhl) [reply hathor/cli/run_node.py]: Done

File: hathor/p2p/manager.py (hunk 10 above)
Comment (jansegre) [reply hathor/p2p/manager.py:614]: While it was added to the built-in `typing` in Python 3.11, it is available in `typing_extensions` for Python 3.10 too (as a backport), we've been using it already.

**Review by jansegre — APPROVED** [review]

File: hathor/cli/run_node.py (hunk 5 above)
Comment (msbrogli) [reply hathor/cli/run_node.py]: Thanks!

File: hathor/metrics.py
Comment (msbrogli) [reply hathor/metrics.py:1]: Thanks!

File: hathor/p2p/peer.py (hunk 11 above)
Comment (msbrogli) [reply hathor/p2p/peer.py]: Thanks!

File: hathor/p2p/peer.py (hunk 12 above)
Comment (msbrogli) [reply hathor/p2p/peer.py]: Thanks!

File: hathor/cli/run_node.py (hunk 4 above)
Comment (luislhl) [reply hathor/cli/run_node.py:220]: Done

Comment (msbrogli) [reply hathor/cli/run_node.py:220]: Thanks!

File: hathor/p2p/manager.py (hunk 9 above)
Comment (msbrogli) [reply hathor/p2p/manager.py:588]: Got it. I guess the following implementation could be more straightforward:

1. Create an enum of types (IPv4=1 and IPv6=2).
2. Create a list of supported addresses (IPv4, IPv6 ou IPv4 | IPv6).
3. Filter all valid addresses. If empty, skip it.

I guess it would prevent some of the most complex logic you are adding.

File: hathor/cli/nginx_config.py (hunk 3 above)
Comment (msbrogli) [reply hathor/cli/nginx_config.py:248]: I agree that there's no harm in enabling it. Let's keep it.

File: hathor/p2p/entrypoint.py (hunk 7 above)
Comment (msbrogli) [reply hathor/p2p/entrypoint.py]: Thanks!

File: hathor/p2p/entrypoint.py (hunk 8 above)
Comment (msbrogli) [reply hathor/p2p/entrypoint.py]: Got it. I agree with you. The current implementation seems good to me.

File: hathor/p2p/manager.py (hunk 10 above)
Comment (msbrogli) [reply hathor/p2p/manager.py:611]: Got it. Make sense. You can keep raising the exception.

**Review by msbrogli — APPROVED** [review]

File: hathor/p2p/manager.py (hunk 9 above)
Comment (jansegre) [reply hathor/p2p/manager.py:588]: > Got it. I guess the following implementation could be more straightforward:
> 
>     1. Create an enum of types (IPv4=1 and IPv6=2).
> 
>     2. Create a list of supported addresses (IPv4, IPv6 ou IPv4 | IPv6).
> 
>     3. Filter all valid addresses. If empty, skip it.
> 
> 
> I guess it would prevent some of the most complex logic you are adding.

This is what is done for `sync-v1` / `sync-v2`. I think it's a good strategy, but I'm OK doing the refactor in a new PR.

**Review by jansegre — APPROVED** [review]

File: hathor/p2p/manager.py (hunk 9 above)
Comment (luislhl) [reply hathor/p2p/manager.py:588]: Are you ok with doing this in another PR, @msbrogli ? 

Then I'll merge this one already

Comment (msbrogli) [reply hathor/p2p/manager.py:588]: Ok. Let's do it separately.
